recipe-server/client/actions/show-heartbeat/index.js

```diff
@@ -16,35 +19,218 @@ export default class ShowHeartbeatAction extends Action {
+    return typeof lastShown !== 'undefined' ?
+      parseFloat(lastShown) : null;
+  }
+
+  /**
    * Checks when this survey was last shown,
    * and returns a boolean indicating if the
-   * user has seen this survey already or not.
+   * user has ever seen this survey or not.
+   *
+   * @async
+   * @return {Boolean}  Has the survey ever been shown?
    */
-  async surveyHasShown() {
-    const lastShown = await this.storage.getItem('lastShown');
+  async hasShownBefore() {
+    const lastShown = await this.getLastShown();
     // If no survey has been shown, lastShown will be falsey.
     return !!lastShown;
   }
 
+  /**
+   * Determines if this heartbeat was shown
+   * at least x days ago.
+   *
+   * @param  {Number}  days Days ago to check
+   * @return {boolean}      Has prompt been shown by that date?
+   */
+  async shownAtleastDaysAgo(days) {
+    const hasShown = await this.hasShownBefore();
+
+    if (!hasShown) {
+      return false;
+    }
+
+    // get timestamp of last shown
+    const timeLastShown = await this.getLastShown();
+
+    // get the difference between now and then
+    const timeElapsed = Date.now() - timeLastShown;
+
+    // time limit is the number of days passed in
+    // converted into milliseconds
+    const timeLimit = ONE_DAY * days;
+
+    // if the diff is smaller than the limit,
+    // that means that the last time the user saw the prompt
+    // was less than the `days` passed in
+    return timeElapsed < timeLimit;
+  }
+
+  /**
+   * Simple function to read the lastInteraction
+   * timestamp (if any) from local storage.
+   * @return {number}   Timestamp of last prompt interaction (if any)
+   */
+  async getLastInteraction() {
+    const lastInteraction = await this.storage.getItem('lastInteraction');
+
+    return typeof lastInteraction !== 'undefined' ?
+      parseFloat(lastInteraction) : null;
+  }
+
+  /**
+   * Gets the timestamp of the last prompt interaction,
+   * and returns the time (in ms) since then.
+   *
+   * @async
+   * @return {number}
+   */
+  async sinceLastInteraction() {
+    const lastInteraction = await this.getLastInteraction();
+
+    return typeof lastInteraction !== 'undefined' ?
+      Date.now() - lastInteraction : null;
+  }
+
+  /**
+   * Checks when the survey prompt last had
+   * interaction from the user (if ever),
+   * and returns a boolean indicating if the
+   * user has ever had interaction
+   *
+   * @async
+   * @return {Boolean}  Has the survey ever had interaction?
+   */
+  async hasHadInteraction() {
+    const lastInteraction = await this.getLastInteraction();
+    return !!lastInteraction;
+  }
+
+  /**
+   * Checks the repeat argument for this recipe,
+   * then determines if the recipe can be qualified as 'ran'.
+   * This ultimately decides if the prompt is shown at all
+   * to the end user.
+   *
+   * @return {boolean}        Has the heartbeat been shown?
+   */
+  async heartbeatHasRan() {
```

Comment:
I rewrote a bunch of the language and ended up renaming the function all together - https://gist.github.com/andymikulski/f2531bdab2bec531ffc0e188f1b62fc2 - Let me know if it looks good and I'll push the changes before merging the branch.